Implement system accounts

abbra opened the pull request:
This PR adds support for system accounts to make easier external management of IPA users.

## Summary by Sourcery

Introduce support for LDAP-based system accounts by adding a dedicated sysaccount plugin with full CLI commands, extending role membership to include system accounts, and providing warnings for passsync configuration updates across servers, along with corresponding documentation and tests.

New Features:
- Implement system accounts LDAP object with CLI commands for add, delete, modify, find, show, enable, and disable.

Enhancements:
- Extend role and baseldap plugins to support sysaccount membership and manage passsync managers with a ServerPasssyncMgrUpdateRequired warning.

Documentation:
- Add a design document for system accounts and regenerate API docs for sysaccount commands and role membership.

Tests:
- Add XMLRPC tests for the sysaccount plugin and update existing role and service plugin tests to cover sysaccount membership.

t-woerner commented:
> @t-woerner this change would be a good candidate as a RFE for ansible-freeipa. It would be nice to sync its development with FreeIPA's release of the feature.

@rjeffman, @abbra Yes this should be coordinated.

abbra commented on API.txt:
```diff
@@ -6027,6 +6029,87 @@ option: Str('version?')
 output: PrimaryKey('value')
+command: sysaccount_add/1
+args: 1,9,3
+arg: Str('uid', cli_name='id')
```
I set it to `login` now but it doesn't matter because it is an argument, not an option and those are never specified by the option name.

```
# ipa help sysaccount-add|head -1
Usage: ipa [global-options] sysaccount-add ID [options]
... [ change to cli_name="login" ] ...
# ipa help sysaccount-add|head -1
Usage: ipa [global-options] sysaccount-add LOGIN [options]
```

abbra replied:
this is fixed, resolving.

t-woerner commented on doc/designs/sysaccounts.md:
```diff
@@ -0,0 +1,184 @@
+these applications require communication with other IPA servers, an update must be executed against
+those servers.
+
+As IPA command line tool, `ipa`, does not currently provide an option to choose which server to run
```
This applies to client context used on an IPA client, but not for an IPA server (client or server context), right?

t-woerner replied:
Wouldn't it be better to limit the modification of system accounts to servers, so that it is clear where it was done?

abbra replied:
A user who executes either `sysaccount-add` or `sysaccount-mod` with `--update-without-reset` option (regardless of the value) will get a message back that tells him about the server the operation was performed on, that the LDAP server on it must be restarted, and that if required, the operations must be performed on other servers as well. This is part of the code already.

abbra commented on doc/designs/sysaccounts.md:
```diff
@@ -0,0 +1,184 @@
+`my-role` role.
+
+Due to the fact that `passSyncManagersDNs` configuration setting is not replicated across IPA
+servers, the `sysaccount-mod` command execution will issue a warning to make sure to update other
```
`-mod` and `-add`. When `-del` is run, it will remove the account from the `ipa_pwd_extop` configuration. I'll update.

`-enable` and `-disable` only touch the ability to auth with the sysaccount, not its properties.

abbra commented:
I renamed `sysaccount-allow-reset` to `sysaccount-manage-reset` command and changed `--update-password-reset=TRUE|FALSE` option to `--privileged=TRUE|FALSE` to simplify it.

abbra commented:
The test passed:
```
[2025-09-01 10:52:44] test_xmlrpc/test_sysaccount_plugin.py::test_sysaccount::test_sysaccount_use . [PASSED]
```

abbra commented:
Successful temp execution report: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/dfd264d4-952d-11f0-8ab9-fa163ed57a39/report.html?sort=result

abbra commented:
FYI, `test_integration-test_installation.py-TestInstallMaster-test_admin_root_alias_CVE_2020_10747` test suite failed because we need to update the F42 image to include 389-ds-base 3.1.3-2 or later.

abbra commented:
The issue with `sysaccount-mod  --privileged=..` is that we don't modify the object itself but `LDAPUpdate` performs modification anyway and re-raises EmptyModList exception. This is the same behavior in few other operations too. Maybe we should add a class variable that prevents this behavior.

Something like
```
class LDAPUpdate(LDAPQuery, crud.Update):
...
     ignore_empty_update = False
...
     def execute(self, *keys, **options):
...
        except errors.EmptyModlist as e:
            if not rdnupdate or not self.ignore_empty_update:
                raise e
          
```

flo-renaud commented on install/ui/src/freeipa/sysaccount.js:
```diff
@@ -0,0 +1,96 @@
+                    sections: [
+                        {
+                            name: "details",
+                            fields: ["uid", "description"],
```
Remove the "description"

abbra commented on ipaserver/plugins/sysaccounts.py:
```diff
@@ -0,0 +1,438 @@
+
+    takes_options = LDAPUpdate.takes_options + update_without_reset
+    has_output_params = LDAPUpdate.has_output_params + update_without_reset
+    allow_empty_update = True
```
I fixed this, @t-woerner already reported non-idempotent behavior.

flo-renaud commented:
Thanks for the update, works for me.
Ack when CI is green
